### ISSUE TITLE
Implement Proportion_Strings sonification

### ISSUE THREAD

**Issue by emily-phet** [phetsims/ratio-and-proportion#45]
See issue #9, particularly this comment https://github.com/phetsims/ratio-and-proportion/issues/9#issuecomment-620843894.

It would be ideal to have this ready for discussion in Tuesday afternoon's sound design meeting.

(assigning @jbphet and @Ashton-Morris as a heads up that @zepumph will be working on this).

**Comment by emily-phet** [comment]
Adding on to this idea:

In Tuesday's meeting, I'd like to discuss potential secondary sounds, such as boundary sounds, and gridline-crossing sounds.

I have the idea we might treat the vertical movement of the hand markers as sliders. Thinking of them that way, boundary sounds and slight 'bumps' as you pass gridlines make a lot of sense! If we can add in the same slider sounds we just refined in Waves Intro, alone and with the Proportion_Strings sound, to get a feel for it, I think it would help the sound design discussion progress.

So - what would be ideal to have by Tuesday's sound design meeting is the following:

In addition to the current sound options available:

- [ ] Add an option for Proportion_Strings, as in @Ashton-Morris demo video in #9 
-[ ] Add an option to hear no sonification (this will allow you to hear only what's listed next in this list...]
-[ ] A new set of options for "Slider" sounds, either On or Off. If Slider sounds is "on" you'll hear the following: boundary sounds (from current slider in Waves Intro) played whenever the top or bottom is reached (regardless of grid selection), and when the grid selection ***does not*** equal "none", we hear the slider bump sounds as each gridline is reached. When "Slider sounds" is "off" you won't hear these. This will allow us to hear these sounds alone, and with the other sonifications, such as "Proportion_Strings", while also being able to hear those other sonifications alone.

**Comment by jbphet** [comment]
Thanks for the heads up, I'll consider myself "on call" for help with the implementation.

@zepumph - there are boundary sounds that are common and that were used in GFL if that helps.

**Comment by zepumph** [comment]
For the first checkbox, I think copying @Ashton-Morris's comment over here will be helpful:

>@zepumph I have some files for you [which are all stored here](https://drive.google.com/open?id=10-kHZgFSu0HEMUfNk0zEn0cvPRK8gFXg). Some of these are .wav so we can have better loops during the prototype phase.

>I added a staccato sine but realistically that's something we'd need to do in web audio programmatically. 

>applause-loop.wav
staccato-sine.mp3
strings-loop-c5.wav
cello-loop-c4.wav
synth-strings-loop-c4.wav
bowed-piano-loop-c4.wav
cello-tremolo-loop-c4.wav

**Comment by zepumph** [comment]
I implemented a first pass of this above. Basically I took the vibrato sound and used it also for this sound option, then I added the `strings-loop-c5.wav` when both hands had enough velocity. The output volume of the strings is mapped from 0-.7 based on the fitness of the proportion.

As to the definition of "enough velocity": Basically I recalculate each velocity every 30 frames (~twice a second), and then if both velocities are greater than |.01| (again each value only ranges from 0-1) then the strings can play.

So to recap, the strings algorithm, in english is like so: Play the strings if both hands are moving more than .01 per .5 seconds, and play it at a volume based on the proportion fitness (from 0-.7).

I'm excited to discuss further tomorrow on potential other strategies. Let me know if anyone would like some other variations in the sim in the next few hours and I can work on it in the morning.

I'll get a dev version out shortly

**Comment by zepumph** [comment]
https://phet-dev.colorado.edu/html/ratio-and-proportion/1.0.0-dev.18/phet/ratio-and-proportion_en_phet.html

**Comment by zepumph** [comment]
The first comment that comes to my mind is that the tolerance for  a "good" proportion is pretty small.  As an educated adult, I find the task of continuously producing strings sounds pretty challenging. We should take this into consideration when determining a general tolerance level in https://github.com/phetsims/ratio-and-proportion/issues/14.

I would hope that we could use the same window for all success feedback (sound/description/color/etc). Developers, you can adjust that in your local copy to play around with it more by changing the value of `FreeObjectsModel.tolerenceProperty`

**Comment by emily-phet** [comment]
Thanks @zepumph! Heads up to @BLFiedler that there's a new dev link above.

**Comment by BLFiedler** [comment]
Just a few quick notes so that I don't forget:
- With just this addition, this feels REALLY close to a complete experience from a touch perspective. Absolutely fantastic.
- It's possible to be moving and have the String sound cut in instead of fade-in, which is a bit jarring. Will need to play more to figure out exactly why this happens.
- The string sound plays when the reset button is pressed (my guess is some weird interaction with how the calculation is done)
- The string sound plays when the two hands are moving in opposing directions. We should consider if this is desired behavior or confusing. My gut reaction is to say they should only play when both hands are increasing in height or both decreasing in height. The other case doesn't strike me as pedagogically relevant and potentially more confusing. Happy to be told I'm wrong though.
- @zepumph You're right on the tolerance. This is something we were going to bring up soon (and I actually meant to add to #14 ). I think we need to open up the tolerance, but not THAT much. I think a small tweak will go far. We should maybe be thinking about the tolerance wrt to grid increments, similar to how we're thinking about alternative input in #44, because they will need to play nicely together.

**Comment by zepumph** [comment]
>@zepumph You're right on the tolerance. This is something we were going to bring up soon (and I actually meant to add to #14 ). I think we need to open up the tolerance, but not THAT much. I think a small tweak will go far. We should maybe be thinking about the tolerance wrt to grid increments, similar to how we're thinking about alternative input in #44, because they will need to play nicely together.

I added a tolerance query parameter over in #14 to help us tweak that.

**Comment by zepumph** [comment]
Feedback from today's strings discussion:

- [ ] have a fade in time, independent of the goal
- [ ] We want the velocities to be matching directions.
- [ ] Fix reset case

The above will be good to do, as they may apply to any sound we implement for the velocity success case.

Next week we plan to listen to a bunch of different sounds to try to find a more pleasant overall experience with our sound mapping. Here are some thoughts we want to look into when finding sounds next week.

Design questions/goals:
* Is there always sound playing while interacting with a ratio half
* Make a sound/loop that doesn't get old/uncomfortable after X minutes of use
* Find our place on the spectrum between unfit ratios having uncomfortable sounds vs comfortable sounds for fit ratios.

**Comment by zepumph** [comment]
I didn't get to the fade in today, but the other two bugs, along with a startup on was solved above.

**Comment by zepumph** [comment]
I also implemented the ratio UI sounds, including having the pickup and drop sounds match the volume for the new boundary and grid "click" sounds.

@emily-phet the only thing left for me here is to implement the fading for the strings sound and any other aesthetic fine tuning we do from there. Would you like to review all but that now?

**Comment by emily-phet** [comment]
@zepumph The latest seems good to go for now, for our upcoming Tuesday discussion.

**Comment by emily-phet** [comment]
@zepumph Actually, one small thing. Can you change the tolerance so that's it's a little easier to get and maintain the strings sound? It would be good to get feedback without people struggling to produce the sound. Note, this is not about optimizing the tolerance, just making sure if we ask for feedback that the person isn't hung up on trying to get the strings sound to happen.

**Comment by emily-phet** [comment]
Just a few notes for Tuesday's design meeting.
We're looking for something pleasant, ideally playful, cohesive across all interactions (moving towards/away from target ratio, at ratio, and maintaining ratio consistently in motion). Some dimensions to explore: 
- Continuous vs staccato sound when moving to/from target ratio.
- Continuous, layered, or looping sound for sound associated with maintaining ratio consistently in motion.

**Comment by BLFiedler** [comment]
@zepumph set a query parameter for tolerance in #14 , but if we're choosing something that's a bit easier, tolerance=0.1 works pretty well I think.

> In the above commit, I made `?tolerance=.05` a query parameter, with that as the default. It felt good to mess around with values even up to .2, just to get an idea of what is "too far." Let me know if I can help this design any other ways.

**Comment by zepumph** [comment]
>@zepumph set a query parameter for tolerance in #14 , but if we're choosing something that's a bit easier, tolerance=0.1 works pretty well I think.

Yes indeed! That said, I really like .1 more and more each day, so I made that the default, you can still update it with the QP as desired.

**Comment by zepumph** [comment]
We continued the discussion and next steps over in https://github.com/phetsims/ratio-and-proportion/issues/9#issuecomment-631089157. I think this issue may be ready to close, though I'm not sure.

**Comment by emily-phet** [comment]
Agreed - moving on with with #9 and beyond. Closing.
